pkg/privilege/privileges/cache.go

```diff
@@ -484,6 +499,165 @@ func (p *MySQLPrivilege) LoadAll(ctx sqlexec.SQLExecutor) error {
 	return nil
 }
 
+func findUserAndAllRoles(userList []string, roleGraph map[auth.RoleIdentity]roleGraphEdgesTable) map[string]struct{} {
+	// Including the user list and also their roles
+	all := make(map[string]struct{}, len(userList))
+	queue := make([]string, 0, len(userList))
+
+	// Initialize the queue with the initial user list
+	for _, user := range userList {
+		all[user] = struct{}{}
+		queue = append(queue, user)
+	}
+
+	// Process the queue using BFS
+	for len(queue) > 0 {
+		user := queue[0]
+		queue = queue[1:]
+		for userHost, value := range roleGraph {
+			if userHost.Username == user {
+				for _, role := range value.roleList {
+					if _, ok := all[role.Username]; !ok {
+						all[role.Username] = struct{}{}
+						queue = append(queue, role.Username)
+					}
+				}
+			}
+		}
+	}
+	return all
```

Comment:
Yes, I write it and this is disgusting.
When updating a user's privilege, all the roles of a user also needs to be updated.
And also the roles of the roles ...
So there is a BFS search to recursively get all the data necessary.